File: exercice2/java/src/main/java/fr/devoxx/devops/logs/sql/SparkSQL5.java
package fr.devoxx.devops.logs.sql;

import fr.devoxx.devops.logs.ApacheAccessLog;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.sql.DataFrame;
import org.apache.spark.sql.SQLContext;
import scala.Tuple4;

import java.io.Serializable;

import static java.lang.Double.valueOf;

/* Statistiques sur la taille des requêtes */
public class SparkSQL5 extends SparkSQL {

    public Tuple4<Long, Long, Double, Long> process(JavaRDD<String> rdd, SQLContext sqlContext) {
        JavaRDD<ApacheAccessLog> accessLogs = rdd.map(ApacheAccessLog::parse);
        configure(sqlContext, accessLogs);

        return sqlContext.sql("select count(size), min(size), avg(size), max(size) from ApacheAccessLog")
                .toJavaRDD()
                .map(row -> new Tuple4<>(
                        row.getLong(0),
                        row.getLong(1),
                        valueOf(row.getDouble(2)),
                        row.getLong(3)
                ))
                .first();
    }

    public static void main(String[] args) {
        if (args.length < 1) {
            System.err.println("Usage: " + SparkSQL5.class.getName() + " <file>");
            System.exit(1);
        }
        SparkConf conf = new SparkConf().setAppName(SparkSQL5.class.getName());
        try (JavaSparkContext sc = new JavaSparkContext(conf)) {
            SparkSQL5 sparkSQL5 = new SparkSQL5();
            Tuple4<Long, Long, Double, Long> result = sparkSQL5.process(sc.textFile(args[0]), new SQLContext(sc));
            System.out.println("->" + result._1() + ", " + result._2() + ", " + result._3() + ", " + result._4());
        }
    }
}

File: exercice2/scala/src/main/scala/fr/devoxx/devops/logs/sql/SparkSQL5.scala
package fr.devoxx.devops.logs.sql

import fr.devoxx.devops.logs.ApacheAccessLog
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.SQLContext

/* Statistiques sur la taille des requêtes */
case class SparkSQL5(rdd: RDD[String] , sqlContext: SQLContext) {

  def process: (Long, Long, Double, Long) = {
    val dataFrame = sqlContext.createDataFrame(rdd.map(ApacheAccessLog.parse));
    dataFrame.registerTempTable("ApacheAccessLog");

    sqlContext.sql("select count(size), min(size), avg(size), max(size) from ApacheAccessLog")
      .map(row => (row.getLong(0), row.getLong(1), row.getDouble(2), row.getLong(3)))
      .first
  }
}
